File: src/chipschallenge/tickbehaviors/GliderTickBehavior.java
package chipschallenge.tickbehaviors;

import chipschallenge.Block;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Move.Moves;

/**
 *
 * @author patrik
 */
public class GliderTickBehavior implements BlockTickBehavior {

    private GliderTickBehavior() {
    }
    private static GliderTickBehavior mInstance = null;

    public static synchronized GliderTickBehavior getInstance() {
        if (mInstance == null) {
            mInstance = new GliderTickBehavior();
        }
        return mInstance;
    }

    private boolean move(Block caller, Moves m) throws BlockContainerFullException {
        //caller.setFacing(m);
        return caller.move(m);
    }

    @Override
    public void tick(Block caller) throws BlockContainerFullException {
        Moves m = caller.getFacing();
        switch (m) {
            case UP:
                if (move(caller, Moves.UP) || move(caller, Moves.LEFT) || move(caller, Moves.RIGHT) || move(caller, Moves.DOWN)) {
                    return;
                }
            case RIGHT:
                if (move(caller, Moves.RIGHT) || move(caller, Moves.UP) || move(caller, Moves.DOWN) || move(caller, Moves.LEFT)) {
                    return;
                }
            case LEFT:
                if (move(caller, Moves.LEFT) || move(caller, Moves.DOWN) || move(caller, Moves.UP) || move(caller, Moves.RIGHT)) {
                    return;
                }
            case DOWN:
                if (move(caller, Moves.DOWN) || move(caller, Moves.RIGHT) || move(caller, Moves.LEFT) || move(caller, Moves.UP)) {
                    return;
                }
        }

    }
}

File: ChipsChallenge.Shared/Tickbehaviors/GliderTickBehavior.cs
namespace ChipsChallenge.Shared.Tickbehaviors
{
    using Moves = Move.Moves;

    public class GliderTickBehavior : IBlockTickBehavior
    {
        private static GliderTickBehavior instance;

        private GliderTickBehavior()
        {
        }

        public static GliderTickBehavior Instance
        {
            get
            {
                lock (typeof(GliderTickBehavior))
                {
                    return instance ?? (instance = new GliderTickBehavior());
                }
            }
        }

        public virtual void Tick(Block caller)
        {
            Moves m = caller.Facing;
            switch (m)
            {
                case Moves.UP:
                    if (Move(caller, Moves.UP) || Move(caller, Moves.LEFT) || Move(caller, Moves.RIGHT) || Move(caller, Moves.DOWN))
                    {
                    }

                    break;
                case Moves.RIGHT:
                    if (Move(caller, Moves.RIGHT) || Move(caller, Moves.UP) || Move(caller, Moves.DOWN) || Move(caller, Moves.LEFT))
                    {
                    }

                    break;
                case Moves.LEFT:
                    if (Move(caller, Moves.LEFT) || Move(caller, Moves.DOWN) || Move(caller, Moves.UP) || Move(caller, Moves.RIGHT))
                    {
                    }

                    break;
                case Moves.DOWN:
                    if (Move(caller, Moves.DOWN) || Move(caller, Moves.RIGHT) || Move(caller, Moves.LEFT) || Move(caller, Moves.UP))
                    {
                    }

                    break;
            }
        }

        private bool Move(Block caller, Moves m)
        {
            Moves before = caller.Facing;
            bool ret = caller.Move(m);
            if (!ret)
            {
                caller.Facing = before;
            }

            return ret;
        }
    }
}